#Лабораторная работа №3

Создание нейронной сети, выполняющей предсказание непрерывной величины

In [ ]:
%%capture
!wget https://www.dropbox.com/s/bbm6rxqb4bsfl2d/training_data.xlsx
!wget https://www.dropbox.com/s/gjhur7eyzcv265y/test_data.xlsx

In [ ]:
import pandas as pd
training_data = pd.read_excel('training_data.xlsx', usecols=lambda x: 'Unnamed' not in x)
test_data = pd.read_excel('test_data.xlsx', usecols=lambda x: 'Unnamed' not in x)
training_data.head()

price  year  condition  cylinders  odometer title_status transmission  \
0  43900  2016          4          6     43500        clean    automatic   
1  15490  2009          2          8     98131        clean    automatic   
2   2495  2002          2          8    201803        clean    automatic   
3   1300  2000          1          8    170305      rebuilt    automatic   
4  13865  2010          3          8    166062        clean    automatic   

  drive       size        lat        long  weather  
0   4wd  full-size  36.471500  -82.483400     59.0  
1   4wd  full-size  40.468826  -74.281734     52.0  
2   4wd  full-size  42.477134  -82.949564     45.0  
3   4wd  full-size  40.764373  -82.349503     49.0  
4   4wd        NaN  49.210949 -123.114720      NaN

Ниже в таблице представлено описание каждого из 12 полей.

|Название поля 	 |Описание      	                              |Название поля   |Описание                                  |
|:--------------:|:----------------------------------------------:|:--------------:|:----------------------------------------:|
|**price**       |Цена                                            |**transmission**|Коробка передач                           |
|**year**        |Год производства 	                              |**drive**       |Привод                                    |
|**condition**   |Состояние        	                              |**size**        |Полноразмер или нет                       |
|**cylinders**   |Количество цлиндров 	                          |**lat**         |Широта 	                                  |
|**odometer**    |Пробег                                          |**long**        |Долгота  	                              |
|**title_status**|Легальный статус авто  (все документы в наличии)|**weather**     |Среднегодовая температура в городе продажи|


Проверим данные на наличие пропусков и типов переменных

Начнем с проверки общей информации о данных.
Для того чтобы это сделать, нужно вызвать у переменной *training_data* метод **info()**.

In [ ]:
training_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4913 entries, 0 to 4912
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   price         4913 non-null   int64  
 1   year          4913 non-null   int64  
 2   condition     4913 non-null   int64  
 3   cylinders     4913 non-null   int64  
 4   odometer      4913 non-null   int64  
 5   title_status  4913 non-null   object 
 6   transmission  4913 non-null   object 
 7   drive         4651 non-null   object 
 8   size          3825 non-null   object 
 9   lat           4913 non-null   float64
 10  long          4913 non-null   float64
 11  weather       4801 non-null   float64
dtypes: float64(3), int64(5), object(4)
memory usage: 460.7+ KB


Анализируем результата выполнения команды:

* 4913 строк (entries)
* 12 столбцов (Data columns)

В данных присутствует три типа dtypes:
* int64 - целое число  (5 столбцов)
* float64 - дробное число (3 столбца)
* object - не число, обычно текст (4 столбца)

В нашем случае признаки с типом object имеют текстовые значения. 

Цифры в каждой строчке обозначают количество заполненных (*non-null*) значений. Видно, что в данных содержатся пропуски, так как эти цифры не в каждой строчке совпадают с полным числом строк (4913).

Для удобства работы выкинем NaN данные из нашего датасета, применив метод **dropna()** к *training_data*:

In [ ]:
training_data = training_data.dropna()

С помощью метода **_get_numeric_data()** оставим столбцы, содержащие только числовые данные.

In [ ]:
training_data = training_data._get_numeric_data()

In [ ]:
training_data.head()

price  year  condition  cylinders  odometer        lat        long  weather
0  43900  2016          4          6     43500  36.471500  -82.483400     59.0
1  15490  2009          2          8     98131  40.468826  -74.281734     52.0
2   2495  2002          2          8    201803  42.477134  -82.949564     45.0
3   1300  2000          1          8    170305  40.764373  -82.349503     49.0
5   6995  2003          3          8    167662  45.518031 -122.578752     50.0

In [ ]:
target_variable_name = 'price'
training_values = training_data[target_variable_name]
training_points = training_data.drop(target_variable_name, axis=1)

Таким образом мы разделили целевую переменную и входные признаки в различные **dataframe**. Аналогичным образом необходимо провести обработку тестовых данных.

В качестве примера создания использовать [пример с лекции](https://colab.research.google.com/drive/1FTaM9W8J88lfygvEDAM1Lk7_9iUhfA2f?usp=sharing).

**Задание (базовый уровень):**
Создать нейронную сеть, которая будет предсказывать цену автомобиля на основе числовых полей исходного датасета (оценку качества регрессора проводить на основе метрик MAE, MSE).

**Задание (повышенный уровень):**
Создать нейронную сеть, которая будет предсказывать цену автомобиля на основе всех полей исходного датасета (оценку качества регрессора проводить на основе метрик MAE, MSE). Для кодирования текстовых полей можно воспользоваться **LabelEncoder** либо **OneHotEncoder** из библиотеки **sklearn.preprocessing**.